File: Leetcode-162_Find Peak Element.cpp
// Time Complexity : O(LogN)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

/*
1. Use binary search to reduce search space.
2. compare the mid with its neighbouring elements to see if its the peak element else goto 3.
3. check the mid-1 and mid+1 and narrow done search to subarray in which elements are in increasing order.

*/

class Solution {
    public:
        int findPeakElement(vector<int>& nums) 
        {
            int low =0, high = nums.size()-1;
            
            int mid;
            while(low<=high)
            {
                mid = low + (high-low)/2;
                
				// mid is the peak ele or when mid is first or last ele considering -infinity on both ends
                if((mid == 0 || nums[mid]>nums[mid-1])&& (mid == nums.size()-1 || nums[mid]> nums[mid+1]))
                   return mid;
                    
					// narrow down search space to the monotonically increasing side.
                else if(nums[mid+1]>nums[mid])
                    low = mid+1;
                    
                else
                    high = mid-1;
            }
                    
            return mid;        
        }
};

File: Leetcode_162_FindPeakElement.py
#Time complexity: O(log n)
#Space Complexity: O(1)

class Solution:
    def findPeakElement(self, nums) -> int:
        left, right=0,len(nums)-1
        while left<right-1:
            mid=(left+right)//2
            if nums[mid]>=nums[mid+1] and nums[mid]>=nums[mid-1]:
                return mid
            if nums[mid+1]>nums[mid]:
                left=mid+1
            else:
                right=mid-1
        if nums[left]>=nums[right]:
            return left
        return right